src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/whenshouldbeused/InputWhenShouldBeUsedSwitchExpression.java

```diff
@@ -0,0 +1,108 @@
+/*
+WhenShouldBeUsed
+
+*/
+
+//non-compiled with javac: Compilable with Java21
+package com.puppycrawl.tools.checkstyle.checks.coding.whenshouldbeused;
+
+public class InputWhenShouldBeUsedSwitchExpression {
+
+    void test(Object j, int x) {
+        int a = switch (j) {
+            case Point(int _, int y): // violation, 'When should be used instead of a single if*.'
+                if (y == 0) {
+                    yield 0;
+                }
+            case Integer _: if (x == 0)  yield 4;
```

Comment:
Maybe I am a bit confused on this syntax, but if you corrected the violations here, would it be the same code?

`Point` will fall through to the Integer if `y != 0`. There would be no fallthrough on a `when` and wouldn't it not hit `Integer` since we triggered as a point?

It seems if you fixed this code, you would need an extra `Point` without a `when` for the fallthrough?

Also, add or point to examples where there is no fall through and we use `:` on case.

Reply:
> There would be no fallthrough on a when and wouldn't it not hit Integer since we triggered as a point? 

That's what I expected. I though that there would be no fall through to a guarded case label. But this is wrong we fall through to the next case without stopping on `when`. sounds weird to me. 

```
PS D:\CS\test\src> cat Test.java
public class Test {
    void test(Object j, int x) {
        int a = switch (j) {
            case Point(int _, int y):
                if (y == 0) {
                    yield 0;
                }
            case Integer _:
                if (x == 0) {
                    yield 4;
                }
            default: { yield 3; }
        };
        System.out.println(a);
    }
    void test2(Object j, int x) {
        int b = switch (j) {
            case Point(int _, int y):
                if (y == 0) {
                    yield 0;
                }
            case Integer _ when x == 0:
                    yield 4;
            default: { yield 3; }
        };
        System.out.println(b);
    }
    public static void main(String[] args) {
        Test t = new Test();

        t.test(new Point(1, 1), 0); // 4
        t.test2(new Point(1, 1), 0); // 4

        t.test(new Point(1, 1), 1); // 3
        t.test2(new Point(1, 1), 1); // 4
    }

    record Point(int x, int y) { }
}
PS D:\CS\test\src> javac Test.java
PS D:\CS\test\src> java Test
4
4
3
4
```

--- 
after this what do you think on how to proceed with this check do we need to check for fall through now? 

sonarline don't mention how they handle this in their [check](https://rules.sonarsource.com/java/tag/java21/RSPEC-6916/). The examples are with enhanced switch only where we can't have a fall through

Reply:
@nrmancuso please share your thoughts

Reply:
Sounds like we should handle most simple case with -> syntax only

Reply:
@mahfouz72 I recommend downloading and installing Eclipse. Sonarlint comes with or can be downloaded from the marketplace. I am not near a computer able to test it right now.

Reply:
@rnveach sonar lint flags this 

![22](https://github.com/user-attachments/assets/7b7a9648-e03e-492f-892e-6dd67a636ab1)

Reply:
I leave it to you and @nrmancuso if we should restrict check like mentioned at https://github.com/checkstyle/checkstyle/pull/15199#discussion_r1671559164 .

sonarlint only violates the `Integer` and not the `Point`? If you resolve the one violation, does it report anything else?
Also your example in the screenshot is not the same, as you encase the `Integer` case with `{}` but not the example above. Does sonarlint only violate this because of the encapsulation or it does it without it?

Reply:
> sonarlint only violates the Integer and not the Point

Yes because the case label `Point` with has `{}`. I don't get the point of why they violate it only when we have `{}`

>  If you resolve the one violation, does it report anything else?

nope once you move `if` to `when` then nothing reported

>  Does sonarlint only violate this because of the encapsulation

Yes, it doesn't violate any case that has no `{}` . IDK why may be a bug ?

> : Even though you used {} in the screenshot, because this is using :, this is still a fall through right?

Yes this is a fall-through whether we have curly brackets or not

Reply:
Example when we have `{}` with point. fall through still exists because it is `:` not `->`

![33](https://github.com/user-attachments/assets/fc69d2ce-3830-40c1-ba7f-9761c62c93b8)

Reply:
There is no fall through in your example since we are using patterns in a switch expression

Reply:
Can we make a post somewhere for those less familar with the syntax when it is a fallthrough and when it is not?

Reply:
Conceptually, we are simply following exist Java paradigms; expressions yield a single value

 I think we can simplify things by first adding support only for arrow syntax, then creating an issue to add support for expressions in general to keep things as simple as possible.

@rnveach @mahfouz72 what do you think about this?

Reply:
I am good with this. Let's just support `->` only for now. and add support for `:` in a separate issue later because it seems like there are nuances with it that will make us slow.

Reply:
> There is no fall through in your example since we are using patterns in a switch expression

actually, [this comment](https://github.com/checkstyle/checkstyle/pull/15199#discussion_r1671306450) shows that there is a fall-through

Reply:
> first adding support only for arrow syntax

I am good with this too.

Reply:
done